### ISSUE TITLE
Piecewise in roots

### ISSUE THREAD

**Issue by proy87** [sympy/sympy#21561]
`roots(x**4 - 11*x**3 + x**2*(1675 - 588*i) + x*(-8414 + 1829*i) + 26692 - 659402*i, x)` gives 4 `Piecewise` roots bu there is one condition under `Piecewise`.

Moreover, when I try to `evalf` a root, I get a very long expression. Only `r.expand().evalf()` gives a simple expression.

**Comment by proy87** [comment]
It is imaginary unit.
`roots(x**4 - 11*x**3 + x**2*(1675 - 588*I) + x*(-8414 + 1829*I) + 26692 - 659402*I, x)` gives

`{Piecewise((11/4 - sqrt(-13037/12 + 392*I - 2*((13037/8 - 588*I)*(4108573/256 - 658819*I)/3 - (5057/8 - 1405*I)**2/8 - (13037/8 - 588*I)**3/108)**(1/3))/2 + sqrt(-13037/6 + 2*((13037/8 - 588*I)*(4108573/256 - 658819*I)/3 - (5057/8 - 1405*I)**2/8 - (13037/8 - 588*I)**3/108)**(1/3) + (5057/4 - 2810*I)/sqrt(-13037/12 + 392*I - 2*((13037/8 - 588*I)*(4108573/256 - 658819*I)/3 - (5057/8 - 1405*I)**2/8 - (13037/8 - 588*I)**3/108)**(1/3)) + 784*I)/2, Eq(4108573/256 - 658819*I + (13037/8 - 588*I)**2/12, 0)), (11/4 - sqrt(-13037/12 - 2*(-4108573/256 - (13037/8 - 588*I)**2/12 + 658819*I)/(3*((13037/8 - 588*I)**3/216 + (5057/8 - 1405*I)**2/16 + sqrt((-4108573/256 - (13037/8 - 588*I)**2/12 + 658819*I)**3/27 + ((13037/8 - 588*I)*(4108573/256 - 658819*I)/3 - (5057/8 - 1405*I)**2/8 - (13037/8 - 588*I)**3/108)**2/4) - (13037/8 - 588*I)*(4108573/256 - 658819*I)/6)**(1/3)) + 392*I + 2*((13037/8 - 588*I)**3/216 + (5057/8 - 1405*I)**2/16 + sqrt((-4108573/256 - (13037/8 - 588*I)**2/12 + 658819*I)**3/27 + ((13037/8 - 588*I)*(4108573/256 - 658819*I)/3 - (5057/8 - 1405*I)**2/8 - (13037/8 - 588*I)**3/108)**2/4) - (13037/8 - 588*I)*(4108573/256 - 658819*I)/6)**(1/3))/2 + sqrt(-13037/6 - 2*((13037/8 - 588*I)**3/216 + (5057/8 - 1405*I)**2/16 + sqrt((-4108573/256 - (13037/8 - 588*I)**2/12 + 658819*I)**3/27 + ((13037/8 - 588*I)*(4108573/256 - 658819*I)/3 - (5057/8 - 1405*I)**2/8 - (13037/8 - 588*I)**3/108)**2/4) - (13037/8 - 588*I)*(4108573/256 - 658819*I)/6)**(1/3) + (5057/4 - 2810*I)/sqrt(-13037/12 - 2*(-4108573/256 - (13037/8 - 588*I)**2/12 + 658819*I)/(3*((13037/8 - 588*I)**3/216 + (5057/8 - 1405*I)**2/16 + sqrt((-4108573/256 - (13037/8 - 588*I)**2/12 + 658819*I)**3/27 + ((13037/8 - 588*I)*(4108573/256 - 658819*I)/3 - (5057/8 - 1405*I)**2/8 - (13037/8 - 588*I)**3/108)**2/4) - (13037/8 - 588*I)*(4108573/256 - 658819*I)/6)**(1/3)) + 392*I + 2*((13037/8 - 588*I)**3/216 + (5057/8 - 1405*I)**2/16 + sqrt((-4108573/256 - (13037/8 - 588*I)**2/12 + 658819*I)**3/27 + ((13037/8 - 588*I)*(4108573/256 - 658819*I)/3 - (5057/8 - 1405*I)**2/8 - (13037/8 - 588*I)**3/108)**2/4) - (13037/8 - 588*I)*(4108573/256 - 658819*I)/6)**(1/3)) + 784*I + 2*(-4108573/256 - (13037/8 - 588*I)**2/12 + 658819*I)/(3*((13037/8 - 588*I)**3/216 + (5057/8 - 1405*I)**2/16 + sqrt((-4108573/256 - (13037/8 - 588*I)**2/12 + 658819*I)**3/27 + ((13037/8 - 588*I)*(4108573/256 - 658819*I)/3 - (5057/8 - 1405*I)**2/8 - (13037/8 - 588*I)**3/108)**2/4) - (13037/8 - 588*I)*(4108573/256 - 658819*I)/6)**(1/3)))/2, True)): 1, Piecewise((11/4 - sqrt(-13037/12 + 392*I - 2*((13037/8 - 588*I)*(4108573/256 - 658819*I)/3 - (5057/8 - 1405*I)**2/8 - (13037/8 - 588*I)**3/108)**(1/3))/2 - sqrt(-13037/6 + 2*((13037/8 - 588*I)*(4108573/256 - 658819*I)/3 - (5057/8 - 1405*I)**2/8 - (13037/8 - 588*I)**3/108)**(1/3) + (5057/4 - 2810*I)/sqrt(-13037/12 + 392*I - 2*((13037/8 - 588*I)*(4108573/256 - 658819*I)/3 - (5057/8 - 1405*I)**2/8 - (13037/8 - 588*I)**3/108)**(1/3)) + 784*I)/2, Eq(4108573/256 - 658819*I + (13037/8 - 588*I)**2/12, 0)), (11/4 - sqrt(-13037/6 - 2*((13037/8 - 588*I)**3/216 + (5057/8 - 1405*I)**2/16 + sqrt((-4108573/256 - (13037/8 - 588*I)**2/12 + 658819*I)**3/27 + ((13037/8 - 588*I)*(4108573/256 - 658819*I)/3 - (5057/8 - 1405*I)**2/8 - (13037/8 - 588*I)**3/108)**2/4) - (13037/8 - 588*I)*(4108573/256 - 658819*I)/6)**(1/3) + (5057/4 - 2810*I)/sqrt(-13037/12 - 2*(-4108573/256 - (13037/8 - 588*I)**2/12 + 658819*I)/(3*((13037/8 - 588*I)**3/216 + (5057/8 - 1405*I)**2/16 + sqrt((-4108573/256 - (13037/8 - 588*I)**2/12 + 658819*I)**3/27 + ((13037/8 - 588*I)*(4108573/256 - 658819*I)/3 - (5057/8 - 1405*I)**2/8 - (13037/8 - 588*I)**3/108)**2/4) - (13037/8 - 588*I)*(4108573/256 - 658819*I)/6)**(1/3)) + 392*I + 2*((13037/8 - 588*I)**3/216 + (5057/8 - 1405*I)**2/16 + sqrt((-4108573/256 - (13037/8 - 588*I)**2/12 + 658819*I)**3/27 + ((13037/8 - 588*I)*(4108573/256 - 658819*I)/3 - (5057/8 - 1405*I)**2/8 - (13037/8 - 588*I)**3/108)**2/4) - (13037/8 - 588*I)*(4108573/256 - 658819*I)/6)**(1/3)) + 784*I + 2*(-4108573/256 - (13037/8 - 588*I)**2/12 + 658819*I)/(3*((13037/8 - 588*I)**3/216 + (5057/8 - 1405*I)**2/16 + sqrt((-4108573/256 - (13037/8 - 588*I)**2/12 + 658819*I)**3/27 + ((13037/8 - 588*I)*(4108573/256 - 658819*I)/3 - (5057/8 - 1405*I)**2/8 - (13037/8 - 588*I)**3/108)**2/4) - (13037/8 - 588*I)*(4108573/256 - 658819*I)/6)**(1/3)))/2 - sqrt(-13037/12 - 2*(-4108573/256 - (13037/8 - 588*I)**2/12 + 658819*I)/(3*((13037/8 - 588*I)**3/216 + (5057/8 - 1405*I)**2/16 + sqrt((-4108573/256 - (13037/8 - 588*I)**2/12 + 658819*I)**3/27 + ((13037/8 - 588*I)*(4108573/256 - 658819*I)/3 - (5057/8 - 1405*I)**2/8 - (13037/8 - 588*I)**3/108)**2/4) - (13037/8 - 588*I)*(4108573/256 - 658819*I)/6)**(1/3)) + 392*I + 2*((13037/8 - 588*I)**3/216 + (5057/8 - 1405*I)**2/16 + sqrt((-4108573/256 - (13037/8 - 588*I)**2/12 + 658819*I)**3/27 + ((13037/8 - 588*I)*(4108573/256 - 658819*I)/3 - (5057/8 - 1405*I)**2/8 - (13037/8 - 588*I)**3/108)**2/4) - (13037/8 - 588*I)*(4108573/256 - 658819*I)/6)**(1/3))/2, True)): 1, Piecewise((11/4 + sqrt(-13037/6 + 2*((13037/8 - 588*I)*(4108573/256 - 658819*I)/3 - (5057/8 - 1405*I)**2/8 - (13037/8 - 588*I)**3/108)**(1/3) - (5057/4 - 2810*I)/sqrt(-13037/12 + 392*I - 2*((13037/8 - 588*I)*(4108573/256 - 658819*I)/3 - (5057/8 - 1405*I)**2/8 - (13037/8 - 588*I)**3/108)**(1/3)) + 784*I)/2 + sqrt(-13037/12 + 392*I - 2*((13037/8 - 588*I)*(4108573/256 - 658819*I)/3 - (5057/8 - 1405*I)**2/8 - (13037/8 - 588*I)**3/108)**(1/3))/2, Eq(4108573/256 - 658819*I + (13037/8 - 588*I)**2/12, 0)), (11/4 + sqrt(-13037/12 - 2*(-4108573/256 - (13037/8 - 588*I)**2/12 + 658819*I)/(3*((13037/8 - 588*I)**3/216 + (5057/8 - 1405*I)**2/16 + sqrt((-4108573/256 - (13037/8 - 588*I)**2/12 + 658819*I)**3/27 + ((13037/8 - 588*I)*(4108573/256 - 658819*I)/3 - (5057/8 - 1405*I)**2/8 - (13037/8 - 588*I)**3/108)**2/4) - (13037/8 - 588*I)*(4108573/256 - 658819*I)/6)**(1/3)) + 392*I + 2*((13037/8 - 588*I)**3/216 + (5057/8 - 1405*I)**2/16 + sqrt((-4108573/256 - (13037/8 - 588*I)**2/12 + 658819*I)**3/27 + ((13037/8 - 588*I)*(4108573/256 - 658819*I)/3 - (5057/8 - 1405*I)**2/8 - (13037/8 - 588*I)**3/108)**2/4) - (13037/8 - 588*I)*(4108573/256 - 658819*I)/6)**(1/3))/2 + sqrt(-13037/6 - 2*((13037/8 - 588*I)**3/216 + (5057/8 - 1405*I)**2/16 + sqrt((-4108573/256 - (13037/8 - 588*I)**2/12 + 658819*I)**3/27 + ((13037/8 - 588*I)*(4108573/256 - 658819*I)/3 - (5057/8 - 1405*I)**2/8 - (13037/8 - 588*I)**3/108)**2/4) - (13037/8 - 588*I)*(4108573/256 - 658819*I)/6)**(1/3) - (5057/4 - 2810*I)/sqrt(-13037/12 - 2*(-4108573/256 - (13037/8 - 588*I)**2/12 + 658819*I)/(3*((13037/8 - 588*I)**3/216 + (5057/8 - 1405*I)**2/16 + sqrt((-4108573/256 - (13037/8 - 588*I)**2/12 + 658819*I)**3/27 + ((13037/8 - 588*I)*(4108573/256 - 658819*I)/3 - (5057/8 - 1405*I)**2/8 - (13037/8 - 588*I)**3/108)**2/4) - (13037/8 - 588*I)*(4108573/256 - 658819*I)/6)**(1/3)) + 392*I + 2*((13037/8 - 588*I)**3/216 + (5057/8 - 1405*I)**2/16 + sqrt((-4108573/256 - (13037/8 - 588*I)**2/12 + 658819*I)**3/27 + ((13037/8 - 588*I)*(4108573/256 - 658819*I)/3 - (5057/8 - 1405*I)**2/8 - (13037/8 - 588*I)**3/108)**2/4) - (13037/8 - 588*I)*(4108573/256 - 658819*I)/6)**(1/3)) + 784*I + 2*(-4108573/256 - (13037/8 - 588*I)**2/12 + 658819*I)/(3*((13037/8 - 588*I)**3/216 + (5057/8 - 1405*I)**2/16 + sqrt((-4108573/256 - (13037/8 - 588*I)**2/12 + 658819*I)**3/27 + ((13037/8 - 588*I)*(4108573/256 - 658819*I)/3 - (5057/8 - 1405*I)**2/8 - (13037/8 - 588*I)**3/108)**2/4) - (13037/8 - 588*I)*(4108573/256 - 658819*I)/6)**(1/3)))/2, True)): 1, Piecewise((11/4 - sqrt(-13037/6 + 2*((13037/8 - 588*I)*(4108573/256 - 658819*I)/3 - (5057/8 - 1405*I)**2/8 - (13037/8 - 588*I)**3/108)**(1/3) - (5057/4 - 2810*I)/sqrt(-13037/12 + 392*I - 2*((13037/8 - 588*I)*(4108573/256 - 658819*I)/3 - (5057/8 - 1405*I)**2/8 - (13037/8 - 588*I)**3/108)**(1/3)) + 784*I)/2 + sqrt(-13037/12 + 392*I - 2*((13037/8 - 588*I)*(4108573/256 - 658819*I)/3 - (5057/8 - 1405*I)**2/8 - (13037/8 - 588*I)**3/108)**(1/3))/2, Eq(4108573/256 - 658819*I + (13037/8 - 588*I)**2/12, 0)), (11/4 - sqrt(-13037/6 - 2*((13037/8 - 588*I)**3/216 + (5057/8 - 1405*I)**2/16 + sqrt((-4108573/256 - (13037/8 - 588*I)**2/12 + 658819*I)**3/27 + ((13037/8 - 588*I)*(4108573/256 - 658819*I)/3 - (5057/8 - 1405*I)**2/8 - (13037/8 - 588*I)**3/108)**2/4) - (13037/8 - 588*I)*(4108573/256 - 658819*I)/6)**(1/3) - (5057/4 - 2810*I)/sqrt(-13037/12 - 2*(-4108573/256 - (13037/8 - 588*I)**2/12 + 658819*I)/(3*((13037/8 - 588*I)**3/216 + (5057/8 - 1405*I)**2/16 + sqrt((-4108573/256 - (13037/8 - 588*I)**2/12 + 658819*I)**3/27 + ((13037/8 - 588*I)*(4108573/256 - 658819*I)/3 - (5057/8 - 1405*I)**2/8 - (13037/8 - 588*I)**3/108)**2/4) - (13037/8 - 588*I)*(4108573/256 - 658819*I)/6)**(1/3)) + 392*I + 2*((13037/8 - 588*I)**3/216 + (5057/8 - 1405*I)**2/16 + sqrt((-4108573/256 - (13037/8 - 588*I)**2/12 + 658819*I)**3/27 + ((13037/8 - 588*I)*(4108573/256 - 658819*I)/3 - (5057/8 - 1405*I)**2/8 - (13037/8 - 588*I)**3/108)**2/4) - (13037/8 - 588*I)*(4108573/256 - 658819*I)/6)**(1/3)) + 784*I + 2*(-4108573/256 - (13037/8 - 588*I)**2/12 + 658819*I)/(3*((13037/8 - 588*I)**3/216 + (5057/8 - 1405*I)**2/16 + sqrt((-4108573/256 - (13037/8 - 588*I)**2/12 + 658819*I)**3/27 + ((13037/8 - 588*I)*(4108573/256 - 658819*I)/3 - (5057/8 - 1405*I)**2/8 - (13037/8 - 588*I)**3/108)**2/4) - (13037/8 - 588*I)*(4108573/256 - 658819*I)/6)**(1/3)))/2 + sqrt(-13037/12 - 2*(-4108573/256 - (13037/8 - 588*I)**2/12 + 658819*I)/(3*((13037/8 - 588*I)**3/216 + (5057/8 - 1405*I)**2/16 + sqrt((-4108573/256 - (13037/8 - 588*I)**2/12 + 658819*I)**3/27 + ((13037/8 - 588*I)*(4108573/256 - 658819*I)/3 - (5057/8 - 1405*I)**2/8 - (13037/8 - 588*I)**3/108)**2/4) - (13037/8 - 588*I)*(4108573/256 - 658819*I)/6)**(1/3)) + 392*I + 2*((13037/8 - 588*I)**3/216 + (5057/8 - 1405*I)**2/16 + sqrt((-4108573/256 - (13037/8 - 588*I)**2/12 + 658819*I)**3/27 + ((13037/8 - 588*I)*(4108573/256 - 658819*I)/3 - (5057/8 - 1405*I)**2/8 - (13037/8 - 588*I)**3/108)**2/4) - (13037/8 - 588*I)*(4108573/256 - 658819*I)/6)**(1/3))/2, True)): 1`

Version of SymPy: 1.8

**Comment by smichr** [comment]
The condition of the Piecewise is of the form `Eq(n, 0)`, e.g. `Eq(4108573/256 - 658819*I + (13037/8 - 588*I)**2/12, 0)`. If the assembly of the Piecewise used expansion of the condition, it would resolve. In the unexpanded state sympy cannot tell if it is 0 or not. In expanded form it is `Eq(2502523/12 - 3274089*I/4, 0) -> False`. Inability to resolve this without expansion is related to the (perhaps) broken numerical testing of Relationals. I'll have to see if I can find the issue number.

**Comment by smichr** [comment]
> find the issue number.

#12327 and associated references

**Comment by oscarbenjamin** [comment]
This seems to be fixed on master and I bisected the fix to 34b4723a7c40c527343875328570b170d83675cb from #21276.

**Comment by smichr** [comment]
Ahh - the expansion has now been done
```
            # whether a Piecewise is returned or not
            # depends on knowing p, so try to put
            # in a simple form
            p = _mexpand(p)
```
